### DISCUSSION TITLE
How to callback with slint_interpreter in rust

### DISCUSSION THREAD

**Discussion by oheil** [slint-ui/slint#4061] [General] [answered]
New to Slint. Playing around a bit and tried to add a callback to a UI generated with slint_interpreter.
Starting with the example:
```
use slint_interpreter::{ComponentDefinition, ComponentCompiler, ComponentHandle};

let mut compiler = ComponentCompiler::default();
let definition =
    spin_on::spin_on(compiler.build_from_path("hello.slint"));
slint_interpreter::print_diagnostics(&compiler.diagnostics());
if let Some(definition) = definition {
    let instance = definition.create().unwrap();
    instance.run().unwrap();
}
```
And a `hello.slint` file like:
```
import { Button, VerticalBox } from "std-widgets.slint";

export component Demo {
    callback button-clicked();
    VerticalBox {
        alignment: start;
        Text {
            text: "Hello World!";
            font-size: 2rem;
            horizontal-alignment: center;
        }
        HorizontalLayout {
            alignment: center;
            Button {
                text: "OK";
                clicked => {root.button-clicked();}
            }
        }
    }
}
```
How can I get e.g. a simple
```
print!("Button clicked\n");
```
executed every time when I click the button? How to implement `on_button_clicked`?
How should I read the docs at https://slint.dev/releases/1.3.2/docs/rust/slint_interpreter/ to find it out on my own?
Thanks in advance for any help.

**Comment by tronical** [comment, marked as answer]
You're probably looking for [`set_callback`](https://slint.dev/releases/1.3.2/docs/rust/slint_interpreter/struct.ComponentInstance#method.set_callback). I think you can call it like so: `instance.set_callback("button-clicked", || { println!("...") });`.

You may have to call it before you call `run()`.

**Comment by oheil** [comment]
Thank you, yes, that's probably what I missed to find. Somehow I though it should be in `ComponentDefinition`... I still have to get used with the docs.

**Comment by oheil** [comment]
Thanks again, it's working.
For others who find this question, the following 
```
let instance = definition.create().unwrap();
instance.set_callback("button-clicked", || { println!("Button clicked") });
instance.run().unwrap();
```
did throw some errors because it expects closure parameters and a return value. I am not a rust expert but with some fiddling I found this
```
use slint_interpreter::{ComponentCompiler, Value, ComponentHandle};
...
let instance = definition.create().unwrap();
let _ = instance.set_callback("button-clicked", |_| { println!("Button clicked");Value::from(0) });
instance.run().unwrap();
```
to be fine and all errors and warnings vanished.
